redis/cluster.py

```diff
@@ -0,0 +1,1884 @@
+                    ' {0} of {1} covered...'.format(
+                        len(self.slots_cache), REDIS_CLUSTER_HASH_SLOTS)
+                )
+            else:
+                # The user set require_full_coverage to False.
+                # In case of full coverage requirement in the cluster's Redis
+                # configurations, we will raise an exception. Otherwise, we may
+                # continue with partial coverage.
+                # see Redis Cluster configuration parameters in
+                # https://redis.io/topics/cluster-tutorial
+                if not self._skip_full_coverage_check and \
+                        self.cluster_require_full_coverage(tmp_nodes_cache):
+                    raise RedisClusterException(
+                        'Not all slots are covered but the cluster\'s '
+                        'configuration requires full coverage. Set '
+                        'cluster-require-full-coverage configuration to no on '
+                        'all of the cluster nodes if you wish the cluster to '
+                        'be able to serve without being fully covered.'
+                        ' {0} of {1} covered...'.format(
+                            len(self.slots_cache), REDIS_CLUSTER_HASH_SLOTS)
+                    )
+
+        # Set the tmp variables to the real variables
+        self.nodes_cache = tmp_nodes_cache
+        self.slots_cache = tmp_slots
+        # Populate the startup nodes with all discovered nodes
+        self.populate_startup_nodes(self.nodes_cache.values())
+
+    def close(self):
+        for node in self.nodes_cache.values():
+            if node.redis_connection:
+                node.redis_connection.close()
+
+    def reset(self):
+        if self.read_load_balancer is not None:
+            self.read_load_balancer.reset()
+
+
+class ClusterPubSub(PubSub):
+    """
+    Wrapper for PubSub class.
+
+    IMPORTANT: before using ClusterPubSub, read about the known limitations
+    with pubsub in Cluster mode and learn how to workaround them:
+    https://redis-py-cluster.readthedocs.io/en/stable/pubsub.html
+    """
+
+    def __init__(self, redis_cluster, node=None, host=None, port=None,
+                 **kwargs):
+        """
+        When a pubsub instance is created without specifying a node, a single
+        node will be transparently chosen for the pubsub connection on the
+        first command execution. The node will be determined by:
+         1. Hashing the channel name in the request to find its keyslot
+         2. Selecting a node that handles the keyslot: If read_from_replicas is
+            set to true, a replica can be selected.
+        """
+        self.node = None
+        connection_pool = None
+        if host is not None and port is not None:
+            node = redis_cluster.get_node(host=host, port=port)
+            self.node = node
+        if node is not None:
+            if not isinstance(node, ClusterNode):
```

Comment:
You can just check isinstance, even with node not being None. This will make this more performant.

```
if isinstance(node, ClusterNode) is not ClusterNode:
    raise DataError("'node' must be a ClusterNode")
```